framework/core/vulkan_resource.h

```diff
@@ -18,149 +19,266 @@
-}
+	using Type = vk::RenderPass;
+};
 
 template <>
-constexpr VkObjectType get_object_type(const VkCommandBuffer &handle)
+struct HPPType<VkSampler>
 {
-	return VK_OBJECT_TYPE_COMMAND_BUFFER;
-}
+	using Type = vk::Sampler;
+};
+}        // namespace detail
 
-/// Inherit this for any Vulkan object with a handle of type `THandle`.
+/// Inherit this for any Vulkan object with a handle of type `HPPHandle`.
 ///
-/// This allows the derived class to store a Vulkan handle, and also a pointer to the parent Device.
-/// It also allow for adding debug data to any Vulkan object.
-template <typename THandle, typename Device = vkb::Device>
+/// This allows the derived class to store a Vulkan handle, and also a pointer to the parent vkb::core::Device.
+/// It also allows to set a debug name for any Vulkan object.
+template <vkb::BindingType bindingType, typename Handle>
 class VulkanResource
 {
   public:
-	VulkanResource(THandle handle = VK_NULL_HANDLE, Device *device = nullptr) :
-	    handle{handle}, device{device}
-	{
-	}
+	using DeviceType = typename std::conditional<bindingType == vkb::BindingType::Cpp, vkb::core::HPPDevice, vkb::Device>::type;
+	using ObjectType = typename std::conditional<bindingType == vkb::BindingType::Cpp, vk::ObjectType, VkObjectType>::type;
+
+	VulkanResource(Handle handle = nullptr, DeviceType *device_ = nullptr);
 
 	VulkanResource(const VulkanResource &)            = delete;
 	VulkanResource &operator=(const VulkanResource &) = delete;
 
-	VulkanResource(VulkanResource &&other) :
-	    handle{other.handle}, device{other.device}
-	{
-		set_debug_name(other.debug_name);
+	VulkanResource(VulkanResource &&other);
+	VulkanResource &operator=(VulkanResource &&other);
 
-		other.handle = VK_NULL_HANDLE;
-	}
+	virtual ~VulkanResource() = default;
+
+	const std::string &get_debug_name() const;
+	DeviceType        &get_device();
+	DeviceType const  &get_device() const;
+	Handle            &get_handle();
+	const Handle      &get_handle() const;
+	uint64_t           get_handle_u64() const;
+	ObjectType         get_object_type() const;
+	bool               has_device() const;
+	bool               has_handle() const;
+	void               set_debug_name(const std::string &name);
+	void               set_handle(Handle hdl);
 
-	VulkanResource &operator=(VulkanResource &&other)
+  private:
+	// we always want to store a vk::Handle as a resource, so we have to figure out that type, depending on the BindingType!
+	template <vkb::BindingType bindingType, typename T>
+	struct DetermineResourceType
+	{
+	};
+	template <typename T>
+	struct DetermineResourceType<vkb::BindingType::Cpp, T>
 	{
-		handle = other.handle;
-		device = other.device;
-		set_debug_name(other.debug_name);
+		using Type = T;
+	};
+	template <typename T>
+	struct DetermineResourceType<vkb::BindingType::C, T>
+	{
+		using Type = typename detail::HPPType<T>::Type;
+	};
+	using ResourceType = typename DetermineResourceType<bindingType, Handle>::Type;
```

Comment:
This construct is a little more complex just because it always stores the C++-handle.
But from C-based code, you tell it the C-handle and also get the C-handle back by just calling `get_handle()`.
And from C++-based code, you tell it the C++-handle and also get the C++-handle back by just calling `get_handle()`.
That is, you trade the complexity of this class for the simplicity of the code using it.

Actually, C-based code usually is not interested (and the coder don't even need to know about) C++-based types, and vice versa. That is, you can use this class by consistently use the types and handles you're used to in your code.

(Note: C-based and C++-based here of course means: C-based Vulkan bindings vs. C++-based Vulkan bindings. Of course, we all code in C++ here!)